------------------------------------------------------------------------------------------------------------
# Assignment 2B
------------------------------------------------------------------------------------------------------------

To Rewrite the Backpropagation table with new random values. The Random values need to be generated with Python code.

#### A NEURON/PERCEPTRON
A NEURON/PERCEPTRON is a special block which can take multiple inputs and produces one output. 
Each inputs is multiplied with a weight to give importance to an input.


#### Step 0:  Load Input and Output 
Read input and output

Python code to Read/load Input and Output values.

In [1]:
import numpy as np
# Create an input matrix with 3 rows and 4 coloumns 
X = np.array([[1,0,1,0], [1,0,1,1], [0,1,0,1]])
# Create an output matrix with 3 rows and 4 coloumns 
y = np.array([[1],[1],[0]])

#### Table with input and output values
Input Matrix
|   |   | X |   |
|---|---|---|---|
| 1 | 0 | 1 | 0 |
| 1 | 0 | 1 | 1 |
| 0 | 1 | 0 | 1 |

Output Matrix
| y |
|---|
| 1 |
| 1 |
| 0 |


#### Step 1: Initialize weights and biases with random values 
Python code to generate & Initialize weights and biases with random values

In [2]:
# Generate random values for input layer weights
wh = np.random.rand(4,3)
# Print the input layer weights
print('input layer weights wh:')
print(wh)
# Generate random values for input layer bias
bh = np.random.rand(1,3)
# Print the input layer bias
print('input layer bias bh:')
print(bh)
# Generate random values for Output layer weights
wout = np.random.rand(3,1)
# Print the output layer weights
print('output layer weights wout:')
print(wout)
# Generate random values for Output layer bias
bout = np.random.rand(1,1)
# Print the output layer bias
print('output layer bias bout:')
print(bout)

input layer weights wh:
[[0.33658772 0.41819905 0.76677351]
 [0.73552224 0.80773914 0.70343078]
 [0.33361382 0.10181493 0.17486657]
 [0.88065084 0.19883932 0.62752478]]
input layer bias bh:
[[0.4127476  0.51384118 0.90913379]]
output layer weights wout:
[[0.68237402]
 [0.48781596]
 [0.2770086 ]]
output layer bias bout:
[[0.36911679]]


#### Table with Input and output weights and biases initialized with random values 

|    | wh |    |
|----|----|----|
|0.33658772 | 0.41819905 | 0.76677351 |
|0.73552224 | 0.80773914 | 0.70343078 |
|0.33361382 | 0.10181493 | 0.17486657 |
|0.88065084 | 0.19883932 | 0.62752478 |


|   | bh |   |
|---|----|---|
| 0.4127476 | 0.51384118 | 0.90913379 |


| wout |
|------|
| 0.68237402 |
| 0.48781596 |
| 0.2770086  |


| bout |
|------|
| 0.36911679 |


#### Step 2: Calculate hidden layer input: 

hidden_layer_input = matrix_dot_product(X,wh) + bh

Python code to compute the hidden_layer_input matrix

In [3]:
hidden_layer_input = np.dot(X, wh) + bh
print('hidden_layer_input matrix:')
print(hidden_layer_input)

hidden_layer_input matrix:
[[1.08294914 1.03385516 1.85077387]
 [1.96359998 1.23269448 2.47829865]
 [2.02892067 1.52041964 2.24008936]]


#### hidden_layer_input matrix

|        | hidden_layer_input |        |
|--------|--------------------|--------|
| 1.08294914 | 1.03385516 | 1.85077387 |
| 1.96359998 | 1.23269448 | 2.47829865 |
| 2.02892067 | 1.52041964 | 2.24008936 |


#### Step 3: Perform non-linear transformation on hidden linear input

hiddenlayer_activations = sigmoid(hidden_layer_input)
Python code to compute the hiddenlayer_activations matrix

In [4]:
# Sigmoid Function definition
def sigmoid(inputs):
    #Calculate the sigmoid for the give inputs (array)    
    sigmoid_value = 1 / 1 + np.exp(- inputs)     
    return sigmoid_value

hiddenlayer_activations = sigmoid(hidden_layer_input)
print('hiddenlayer_activations matrix:')
print(hiddenlayer_activations)

hiddenlayer_activations matrix:
[[1.33859549 1.35563329 1.15711553]
 [1.14035225 1.29150606 1.08388582]
 [1.13147735 1.21862013 1.10644899]]


#### hidden layer activations Matrix

|     |  hiddenlayer_activations |     |
|-----|--------------------------|-----|
| 1.33859549 | 1.35563329 | 1.15711553 |
| 1.14035225 | 1.29150606 | 1.08388582 |
| 1.13147735 | 1.21862013 | 1.10644899 |


#### Step 4: Perform linear and non-linear transformation of hidden layer activation at output layer

output_layer_input = matrix_dot_product (hiddenlayer_activations * wout ) + bout 
output = sigmoid(output_layer_input)

Python code to compute the output matrix.

In [5]:
output_layer_input = np.dot(hiddenlayer_activations, wout) + bout
output = sigmoid(output_layer_input)
print('output_layer_input matrix:')
print(output_layer_input)
print('output matrix:')
print(output)

output_layer_input matrix:
[[2.26437007]
 [2.07752649]
 [2.04216577]]
output matrix:
[[1.10389546]
 [1.12523961]
 [1.1297474 ]]


#### output Matrix

| output     |
|------------|
| 1.10389546 |
| 1.12523961 |
| 1.1297474  |


#### Step 5: Calculate gradient of Error(E) at output layer

E = y-output

Python code to compute the Gradient error matrix.

In [6]:
E = y - output
print('Error matrix:')
print(E)

Error matrix:
[[-0.10389546]
 [-0.12523961]
 [-1.1297474 ]]


#### Error(E) Matrix

|      E       |
|--------------|
|  -0.10389546 |
|  -0.12523961 |
| -1.1297474   |


#### Step 6: Compute slope at output and hidden layer

`Slope_output_layer= derivatives_sigmoid(output)

Slope_hidden_layer = derivatives_sigmoid(hiddenlayer_activations)`

Python code to compute the Slope_output_layer and Slope_hidden_layer matrix.


In [14]:
def derivatives_sigmoid(inputs):
    #Calculate the sigmoid for the give inputs (array)    
    sigmoid_derivative_value = inputs * (1-inputs)
    return sigmoid_derivative_value

slope_output = derivatives_sigmoid(output)
slope_hidden_layer = derivatives_sigmoid(hiddenlayer_activations)
print('slope_output matrix:')
print(slope_output)
print('slope_hidden_layer matrix:')
print(slope_hidden_layer)

slope_output matrix:
[[-0.11468973]
 [-0.14092457]
 [-0.14658179]]
slope_hidden_layer matrix:
[[-0.45324239 -0.48210833 -0.18180082]
 [-0.160051   -0.37648185 -0.09092265]
 [-0.14876365 -0.26641489 -0.11778038]]


#### Slope_output and Slope_hidden_layer Matrix

|Slope_output |
|-------------|
| -0.11468973 |
| -0.14092457 |
| -0.14658179 |


|         |Slope_hidden_layer|        |
|-----------|------------|------------|
| -0.45324239 | -0.48210833 | -0.18180082 |
| -0.160051   | -0.37648185 | -0.09092265 |
| -0.14876365 | -0.26641489 | -0.11778038 |

#### Step 7: Compute delta at output layer

d_output = E * slope_output * lr

Python code to compute the d_output matrix.

In [16]:
# learning rate is represented as lr
lr = 1.0
d_output = E * slope_output * lr
print('d_output matrix:')
print(d_output)

d_output matrix:
[[0.01191574]
 [0.01764934]
 [0.1656004 ]]


#### d_output Matrix

| d_output    |
|-------------|
|  0.01191574 |
|  0.01764934 |
|  0.1656004  |

#### Step 8: Calculate Error at hidden layer

Error_at_hidden_layer = matrix_dot_product(d_output, wout.Transpose)

Python code to compute the Error_at_hidden_layer matrix.


In [17]:
error_at_hidden_layer = np.dot(d_output, wout.T)
print('error_at_hidden_layer matrix:')
print(error_at_hidden_layer)

error_at_hidden_layer matrix:
[[0.00839725 0.00609956 0.00355832]
 [0.01243782 0.00903454 0.0052705 ]
 [0.11670175 0.08476937 0.04945212]]


####  error_at_hidden_layer Matrix

|          | error_at_hidden_layer |      |
|-------------|-------------|-------------|
|  0.00839725 | 0.00609956  | 0.00355832  |
|  0.01243782 | 0.00903454  | 0.0052705   |
|  0.11670175 | 0.08476937  | 0.04945212  |


#### Step 9: Compute delta at hidden layer

d_hiddenlayer = Error_at_hidden_layer * slope_hidden_layer

Python code to compute the d_hiddenlayer matrix.

In [18]:
d_hiddenlayer = error_at_hidden_layer * slope_hidden_layer
print('d_hiddenlayer matrix:')
print(d_hiddenlayer)

d_hiddenlayer matrix:
[[-0.00380599 -0.00294065 -0.0006469 ]
 [-0.00199069 -0.00340134 -0.00047921]
 [-0.01736098 -0.02258382 -0.00582449]]


####  d_hiddenlayer Matrix

|             |d_hiddenlayer|              |
|-------------|-------------|--------------|
|  -0.00380599 | -0.00294065 | -0.0006469  |
|  -0.00199069 | -0.00340134 | -0.00047921 |
|  -0.01736098 | -0.02258382 | -0.00582449 |



#### Step 10: Update weight at both output and hidden layer

wout = wout + matrix_dot_product (hiddenlayer_activations.Transpose, d_output) * learning_rate

wh = wh+ matrix_dot_product (X.Transpose,d_hiddenlayer) * learning_rate

Python code to compute the wout and wh matrix.

In [19]:
learning_rate = 0.1
wh = wh + np.dot(X.T, d_hiddenlayer) * learning_rate
wout = wout + np.dot(hiddenlayer_activations.T, d_output) * learning_rate
print('wh matrix:')
print(wh)
print('wout matrix:')
print(wout)

wh matrix:
[[0.33544676 0.41696048 0.76655644]
 [0.73210509 0.80332859 0.70230804]
 [0.33247287 0.10057636 0.1746495 ]
 [0.87684186 0.1937645  0.62630967]]
wout matrix:
[[0.727064  ]
 [0.53596628]
 [0.32023783]]


#### wh and wout matrix:

|             | wh         |            |
|-------------|------------|------------|
|  0.33544676 | 0.41696048 | 0.76655644 |
| 0.73210509  | 0.80332859 | 0.70230804 |
|  0.33247287 | 0.10057636 | 0.1746495  |
|  0.87684186 | 0.1937645  | 0.62630967 |


|    w_out   |
|------------|
| 0.727064   |
| 0.53596628 |
| 0.32023783 |


#### Update biases at both output and hidden layer

bh = bh + sum(d_hiddenlayer, axis=0) * learning_rate
bout = bout + sum(d_output, axis=0)*learning_rate

Python code to compute the bout and bh matrix.

In [20]:
bh = bh + np.sum(d_hiddenlayer, axis=0) * learning_rate
bout = bout + np.sum(d_output, axis=0) * learning_rate
print('bh matrix:')
print(bh)
print('bout matrix:')
print(bout)

bh matrix:
[[0.40818949 0.50819206 0.90779398]]
bout matrix:
[[0.40814988]]


#### bh and bout matrix

|            |    b_out   |            |
|------------|------------|------------|
| 0.40818949 | 0.50819206 | 0.90779398 |


|    b_h     |
|------------|
| 0.40814988 |